### PR TITLE
DAOS-2342 client: Object Key Checksums

### PR DESCRIPTION
**Pull request by ryon-jensen**
On an object update/fetch the client will calculate checksums for
A Keys and D Keys which will be checked by the server to verify that no
corruption occurred to the key in transit. On key enumeration, the server
will calculate checksums for keys and the cliente will verify.

- The I/O Descriptor A Key checksum field was change to be a pointer to
reduce the descriptor size when checksums are not enabled.
- Add dkey checksum to appropriate structures

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

**Comment by ryon-jensen** [comment]
> Why do we have csum fields in iod to begin with? User has no checksum input, right?

No, user has no checksum input and there's a story to remove them from the iod. But for now it's definitely convenient for the internal APIs.

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -2173,6 +2178,10 @@ obj_comp_cb(tse_task_t *task, void *data)
 	case DAOS_OBJ_RPC_QUERY_KEY:
 	case DAOS_OBJ_RPC_SYNC:
 		obj = *((struct dc_object **)data);
+		if (obj_auxi->opc == DAOS_OBJ_RPC_UPDATE ||
+		    obj_auxi->opc == DAOS_OBJ_RPC_FETCH)
+			/** checksums sent, need to destroy now */
+			obj_rw_csum_destroy(obj, dc_task_get_args(task));
```
Comment (liuxuezhao) [inline src/object/cli_obj.c]: not sure why you move it to here?
the checksum calculate need not to do it again for io retry case right? then only need to destroy csum when need not retry IO, or maybe I misunderstood it?

File: src/object/cli_obj.c [hunk 2]
```diff
@@ -2414,6 +2478,10 @@ do_dc_obj_fetch(tse_task_t *task, daos_obj_fetch_t *args,
 	if (rc != 0)
 		D_GOTO(out_task, rc);
 
+	rc = csum_obj_fetch(obj, args);
```
Comment (liuxuezhao) [inline src/object/cli_obj.c]: same as update case that only need to calculate the key csum when non-retry case right? (if (!obj_auxi->io_retry))

File: src/object/obj_rpc.c [hunk 3]
```diff
@@ -344,6 +349,14 @@ crt_proc_daos_iod_t(crt_proc_t proc, crt_proc_op_t proc_op, daos_iod_t *dvi,
 				D_GOTO(free, rc = -DER_NOMEM);
 		}
 	}
+	if (existing_flags & IOD_KCSUM_EXIST) {
+		rc = crt_proc_daos_csum_buf_t(proc, dvi->iod_kcsum);
```
Comment (liuxuezhao) [inline src/object/obj_rpc.c]: this is good, to reduce the RPC size as far as possible. we may need similar method for the added orw_dkey_csum as comment below.

File: src/include/daos_srv/daos_server.h [hunk 4]
```diff
@@ -542,6 +544,7 @@ struct dss_enum_arg {
 			int			kds_cap;
 			int			kds_len;
 			d_sg_list_t	       *sgl;
+			d_iov_t			csum_iov;
```
Comment (liuxuezhao) [inline src/include/daos_srv/daos_server.h]: maybe @wangdi1 can review key csum related change for enumerate, thx

File: src/object/cli_obj.c [hunk 5]
```diff
@@ -2340,26 +2347,83 @@ obj_null_csum(const daos_obj_update_t *args) {
...
 				  iod, &iod->iod_csums);
+		if (rc != 0)
+			return rc;
+
 		if (DAOS_FAIL_CHECK(DAOS_CHECKSUM_UPDATE_FAIL))
 			((char *) iod->iod_csums->cs_csum)[0]++;
+
+		rc = daos_csummer_calc_key(csummer, &iod->iod_name,
+					   &iod->iod_kcsum);
+		if (rc != 0)
```
Comment (Nasf-Fan) [inline src/object/cli_obj.c]: When the buffer that is allocated in above steps can be freed for the failure case?

File: src/iosrv/vos.c [hunk 6]
```diff
@@ -138,14 +138,68 @@ fill_obj(daos_handle_t ih, vos_iter_entry_t *entry, struct dss_enum_arg *arg,
...
+
+	if (rc != 0)
+		return rc;
+
+	/** Make sure the csum buffer is big enough ... resize if needed */
+	if (csum_iov->iov_buf == NULL) {
+		/** This must be freed by the object layer
+		 * (currently in obj_enum_complete)
+		 */
+		D_ALLOC(csum_iov->iov_buf, csum_len);
```
Comment (Nasf-Fan) [inline src/iosrv/vos.c]: check whether D_ALLOC() is successful or not.

**Comment by ryon-jensen** [comment]
> The logic in obj_verify.c also needs to handle the dkey/akey checksum in the enumeration RPC reply.

I looked at obj_verify.c and it wasn't clear to me where key checksums need to be handled. Can you point me to what you're referring to?

**Comment by ryon-jensen** [comment]
I rebased this patch to the API changes that removed checksums from the IOD which addresses several of the comments here. ( Rebase changes history ... sorry - :-8 )

File: src/object/rpc_csum.c [hunk 7]
```diff
@@ -97,22 +150,24 @@ crt_proc_struct_dcs_iod_csums(crt_proc_t proc, struct dcs_iod_csums *iod_csum)
...
 		for (i = 0; i < iod_csum->ic_nr; i++)
-			PROC(struct_dcs_csum_info, &iod_csum->ic_data[i]);
+			PROC_CI(&iod_csum->ic_data[i]);
 	}
 
-	if (FREEING(proc) && iod_csum != NULL) {
+	if (FREEING(proc)) {
 		for (i = 0; i < iod_csum->ic_nr; i++)
-			PROC(struct_dcs_csum_info, &iod_csum->ic_data[i]);
+			PROC_CI(&iod_csum->ic_data[i]);
```
Comment (jolivier23) [inline src/object/rpc_csum.c]: we just leak iod_csum->ic_data?

File: src/object/rpc_csum.c [hunk 8]
```diff
@@ -87,6 +98,48 @@ crt_proc_struct_dcs_csum_info(crt_proc_t proc, struct dcs_csum_info *csum)
...
+		rc = proc_struct_dcs_csum_info(proc, *p_csum);
+		if (rc != 0) {
+			D_FREE(*p_csum);
+			return rc;
+		}
+	}
+
+	if (FREEING(proc_op)) {
+		PROC_CI(*p_csum);
+		D_FREE(*p_csum);
```
Comment (jolivier23) [inline src/object/rpc_csum.c]: does p_csum leak if there is an error?

File: src/object/rpc_csum.c (hunk 7 above)
Comment (ryon-jensen) [reply src/object/rpc_csum.c]: yes, you're right, I'll fix it

File: src/object/rpc_csum.c (hunk 8 above)
Comment (ryon-jensen) [reply src/object/rpc_csum.c]: yes, you're right, I'll fix it

**Comment by ryon-jensen** [comment]
> Did we decide against making the key csum and hash the same thing?

Initially that was our approach, but at some point changed it to actually calculate a separate checksum, though it isn't stored. I can't remember why the change, but this is how it's documented in the wiki.

File: src/object/cli_obj.c [hunk 9]
```diff
@@ -2501,6 +2560,14 @@ do_dc_obj_fetch(tse_task_t *task, daos_obj_fetch_t *args,
 	if (rc != 0)
 		D_GOTO(out_task, rc);
 
+	if (!obj_auxi->io_retry) {
+		rc = csum_obj_fetch(obj, args, obj_auxi);
```
Comment (liuxuezhao) [inline src/object/cli_obj.c]: minor, maybe can return if !io_retry inside csum_obj_fetch to be the same style as obj_rw_bulk_prep. same for csum_obj_update

File: src/object/cli_shard.c [hunk 10]
```diff
@@ -415,6 +415,10 @@ dc_obj_shard_rw(struct dc_obj_shard *shard, enum obj_rpc_opc opc,
 	orw->orw_dkey_hash = args->dkey_hash;
 	orw->orw_nr = nr;
 	orw->orw_dkey = *dkey;
+	if (args->dkey_csum != NULL)
+		orw->orw_dkey_csum = args->dkey_csum;
+	else
+		memset(&orw->orw_dkey_csum, 0, sizeof(orw->orw_dkey_csum));
```
Comment (liuxuezhao) [inline src/object/cli_shard.c]: orw->orw_dkey_csum is a pointer right? then orw->orw_dkey_csum = NULL is enough
actually the orw is zeroed by calloc, seems fine to remove the else case.

File: src/object/obj_rpc.h [hunk 11]
```diff
@@ -158,6 +158,7 @@ struct obj_iod_array {
 	((uint32_t)		(orw_start_shard)	CRT_VAR) \
 	((uint32_t)		(orw_flags)		CRT_VAR) \
 	((daos_key_t)		(orw_dkey)		CRT_VAR) \
+	((struct dcs_csum_info)	(orw_dkey_csum)		CRT_PTR) \
 	((struct obj_iod_array)	(orw_iod_array)		CRT_VAR) \
 	((struct dcs_iod_csums)	(orw_iod_csums)		CRT_ARRAY) \
```
Comment (liuxuezhao) [inline src/object/obj_rpc.h]: BTW, a minor issue that orw_iod_csums in input struct and orw_iod_csum (no 's') in output struct. is it intended?
if no I may rename them to be same in my patch.

Comment (ryon-jensen) [reply src/object/obj_rpc.h]: No, they should be the same. My intention is if there are multiple iods, then the csum field/param name should be iods_csums. If just one iod, then the csum field/param should be iod_csums. IODs can have multiple csums.

**Comment by ryon-jensen** [comment]
If it's alright, I'm going to have a followup patch to address the minor feedback instead of changing this one which would require re-approvals.

**Pull request merged**